gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientWriteChannel.java

```diff
@@ -0,0 +1,199 @@
+
+  private static WriteChannel getClientWriteChannel(
+      Storage storage,
+      StorageResourceId resourceId,
+      CreateObjectOptions createOptions,
+      GoogleCloudStorageOptions storageOptions) {
+    AsyncWriteChannelOptions channelOptions = storageOptions.getWriteChannelOptions();
+    WriteChannel writeChannel =
+        storage.writer(
+            getBlobInfo(resourceId, createOptions),
+            generateWriteOptions(createOptions, channelOptions));
+    writeChannel.setChunkSize(channelOptions.getUploadChunkSize());
+
+    return writeChannel;
+  }
+
+  private class UploadOperation implements Callable<Boolean> {
+
+    // Read end of the pipe.
+    private final InputStream pipeSource;
+    private final StorageResourceId resourceId;
+    private final int MAX_BYTES_PER_MESSAGE = MAX_WRITE_CHUNK_BYTES.getNumber();
+
+    UploadOperation(InputStream pipeSource, StorageResourceId resourceId) {
+      this.resourceId = resourceId;
+      this.pipeSource = pipeSource;
+    }
+
+    @Override
+    public Boolean call() throws Exception {
+      // Try-with-resource will close this end of the pipe so that
+      // the writer at the other end will not hang indefinitely.
+      try (pipeSource) {
+        boolean lastChunk = false;
+        ByteBuffer byteBuffer = ByteBuffer.allocate(MAX_BYTES_PER_MESSAGE);
+        while (!lastChunk) {
+          int remainingCapacity = byteBuffer.remaining();
+          ByteString data =
+              ByteString.readFrom(
+                  ByteStreams.limit(pipeSource, remainingCapacity), remainingCapacity);
+          if (data.size() < remainingCapacity) {
+            lastChunk = true;
+          }
+          byteBuffer.put(data.toByteArray());
+          // switch to read mode
+          byteBuffer.flip();
+          // this could result into partial write
+          writeInternal(byteBuffer);
+          if (!lastChunk) {
+            // compact buffer for further writing
+            byteBuffer.compact();
+          }
+        }
+        // last chunk could be partially written
+        // uploading all bytes of last chunk
+        if (lastChunk && byteBuffer.hasRemaining()) {
+          while (byteBuffer.hasRemaining()) {
+            writeInternal(byteBuffer);
+          }
+        }
+        return true;
+      } catch (Exception e) {
+        throw new IOException(
+            String.format("Error occurred while uploading resource %s", resourceId), e);
+      }
+    }
+  }
+
+  private static BlobWriteOption[] generateWriteOptions(
+      CreateObjectOptions createOptions, AsyncWriteChannelOptions channelOptions) {
+    List<BlobWriteOption> writeOptions = new ArrayList<>();
+
+    writeOptions.add(BlobWriteOption.disableGzipContent());
+    writeOptions.add(BlobWriteOption.generationMatch());
+    if (createOptions.getKmsKeyName() != null) {
+      writeOptions.add(BlobWriteOption.kmsKeyName(createOptions.getKmsKeyName()));
+    }
+    if (channelOptions.isGrpcChecksumsEnabled()) {
```

Comment:
`fs.gs.checksum.type` is used i na totally different context.

`channelOptions. isGrpcChecksumsEnabled`: is the configuration to tell library that grpc checksum is used while uploading files.

`fs.gs.checksum.type`: is the configuration to decide what checksum to return when `getFileChecksum` API is called against a file.

`fs.gs.checksum.type` configuration has nothing to do with checksum while uploading files.

Reply:
Any reason why we are using transport specific configs? i.e. why this applies only to gRPC?

Reply:
Because checksum support varies based upon the transport https://docs.google.com/document/d/1CF2QqdIkh-BYNA4CdBRqRpJoXp16u6-kh-k62zgWPgM/edit#heading=h.wzygpepoukd

Here I have maintained the parity between existing and new implementation.